### DISCUSSION TITLE
Instrumenting classes that are created through Reflection (ie - Spring XML Beans)

### DISCUSSION THREAD

**Discussion by zackman0010** [open-telemetry/opentelemetry-java-instrumentation#8912] [Q&A] [answered]
I'm attempting to instrument a class that is instantiated through Spring XML Beans. However, no matter what I do, the advice I'm attempting to inject is never being injected.
The specific class is `SimpleRemoteStatelessSessionProxyFactoryBean`. However, the method I'm instrumenting is defined on `abstract class AbstractSlsbInvokerInterceptor`, which is a super-class of SRSSPFB.
I've attempted the following typeMatchers:
- `named("org.springframework.ejb.access.AbstractSlsbInvokerInterceptor")`
- `extendsClass(named("org.springframework.ejb.access.AbstractSlsbInvokerInterceptor"))`
- `named("org.springframework.ejb.access.SimpleRemoteStatelessSessionProxyFactoryBean")`

I've tried turning on the debug logging to see if any exceptions are being thrown, but I see nothing. My instrumentation is enabled, because I do see the "Installing instrumentation" debug message.

My suspicion is that, because I'm using Spring XML configuration, SRSSPFB isn't actually directly referenced anywhere. It's instantiated at runtime through reflection. Is the auto-instrumentation unable to instrument classes instantiated this way? And if that is the case, is there a way around it?

**Comment by laurit** [comment, marked as answer]
Most likely the problem is that by default agent excludes some classes from instrumentation. See https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/164358d78827e9b0e8d75a5e38b7f7c9a80571f8/javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ignore/AdditionalLibraryIgnoredTypesConfigurer.java#L58 To get around this implement a `IgnoredTypesConfigurer` like https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/164358d78827e9b0e8d75a5e38b7f7c9a80571f8/instrumentation/internal/internal-application-logger/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/internal/logging/ApplicationLoggingIgnoredTypesConfigurer.java#L18

**Reply by zackman0010** [reply]
That appears to be exactly the reason! Do you know why the entire EJB package was added to the ignore list? I'd like to avoid breaking anything of it was put there for a reason.

**Reply by laurit** [reply]
Probably to reduce startup overhead. Hopefully we can remove it one day.

**Reply by zackman0010** [reply]
Just confirmed that my instrumentation works by allowing `.allowClass("org.springframework.ejb.access.")`.
Can a note about the default ignored types be added into the contributing documentation somewhere? I'm not sure exactly where it would best fit, but I spent all day yesterday trying to figure out why my instrumentation wasn't working when it ended up being something so simple.

**Reply by trask** [reply]
> Can a note about the default ignored types be added into the contributing documentation somewhere?

sure!

> I'm not sure exactly where it would best fit

maybe a section in this doc about `IgnoredTypesConfigurer`? https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation-module.md
